### PR TITLE
Refactor builder.js

### PR DESCRIPTION
**Pull request by inglesp**
This PR refactors builder.js.  It breaks a big component up into lots of smaller ones, and uses props correctly.

### CODE REVIEW

File: static/js/builder.js [hunk 1]
```diff
@@ -1,56 +1,34 @@
...
+      this.state["status-" + code] = props.codeToStatus[code];
       this.state["expanded-" + code] = true;
     });
 
     this.updateStatus = this.updateStatus.bind(this);
     this.toggleVisibility = this.toggleVisibility.bind(this);
+    this.getStatus = this.getStatus.bind(this);
+    this.getHasDescendants = this.getHasDescendants.bind(this);
+    this.getIsExpanded = this.getIsExpanded.bind(this);
+    this.getIsVisible = this.getIsVisible.bind(this);
```
Comment (ghickman) [inline static/js/builder.js:20]: You can avoid this step by doing `someProp={() => this.someMethod()}` at the call site.

Comment (inglesp) [reply static/js/builder.js:20]: Doesn't this create a new function on every render?

File: static/js/builder.js [hunk 2]
```diff
@@ -392,7 +461,21 @@ function getCookie(name) {
   return cookieValue;
 }
 
+function readValueFromPage(id) {
+  return JSON.parse(document.getElementById(id).textContent);
+}
```
[the remark this answers is not in the document]
Comment (inglesp) [reply static/js/builder.js:466]: That's quite neat.

File: static/js/builder.js (hunk 1 above)
Comment (ghickman) [reply static/js/builder.js:20]: [Apparently so](https://reactjs.org/docs/faq-functions.html#arrow-function-in-render), TIL!

**Pull request merged**